ext/page_log/palm/palm_kv.c

```diff
@@ -372,6 +373,110 @@ palm_kv_put_page(PALM_KV_CONTEXT *context, uint64_t table_id, uint64_t page_id,
     return (mdb_put(context->lmdb_txn, context->env->lmdb_pages_dbi, &kval, &vval, 0));
 }
 
+int
+palm_kv_get_page_ids(
+  PALM_KV_CONTEXT *context, WT_ITEM *item, uint64_t checkpoint_lsn, uint64_t table_id, size_t *size)
+{
+    MDB_cursor *cursor;
+    MDB_stat stat;
+    int ret;
+    size_t count = 0;
+    cursor = NULL;
+    MDB_val kval;
+    MDB_val vval;
+    PAGE_KEY page_key;
+
+    memset(&kval, 0, sizeof(kval));
+    memset(&vval, 0, sizeof(vval));
+    memset(&page_key, 0, sizeof(page_key));
+    page_key.table_id = table_id;
+    swap_page_key(&page_key, &page_key);
+    kval.mv_size = sizeof(page_key);
+    kval.mv_data = &page_key;
+
+    if ((ret = mdb_cursor_open(context->lmdb_txn, context->env->lmdb_pages_dbi, &cursor)) != 0)
+        return (ret);
+
+    /* Get maximum size for page ids */
+    if ((ret = mdb_stat(context->lmdb_txn, context->env->lmdb_pages_dbi, &stat)) != 0)
+        return (ret);
+
+    /* If no entries found, return an error  */
+    if (stat.ms_entries == 0) {
+        item->size = 0;
+        item->data = NULL;
+        return (WT_NOTFOUND);
+    }
+
+    if (item != NULL) {
+        memset(item, 0, sizeof(WT_ITEM));
+        if ((ret = palm_resize_item(item, stat.ms_entries * sizeof(uint64_t))) != 0)
+            return (ret);
+    }
+
+    if (item->data == NULL)
+        return (ENOMEM);
+
+    if ((ret = mdb_cursor_get(cursor, &kval, &vval, MDB_SET_RANGE)) != 0)
+        return (ret);
+
+    uint64_t prev_page_id = 0;
+    int prev_is_tombstone = 0;
+
+    /*
+     * Iterate through the pages table, looking for pages that has a lsn smaller than the given
+     * checkpoint lsn. Note that the pages are sorted by table_id, page_id, lsn in ascending order,
+     * so we are only interested in the last page of each page_id. If the last page is a tombstone,
+     * meaning we're discarding it, then we skip it. If the last page is a full page, we store the
+     * page ID in the item->data array.
```

Comment:
the logic below looks like it accounts for deltas, but this comment doesn't explain that.

Reply:
Actually, even if the last page is a delta, we’re still only tracking the last full page, since `prev_page_id` is only recorded for full pages. The re-assignment of `prev_page_id` wouldn’t be executed in the delta case, because it’s skipped earlier with a `continue` when we detect the current page is a delta page, even if it’s the last page. So the behavior remains consistent.

Reply:
Ah that makes sense. I was reading "last" as in last update on the chain but you mean last as in previous. Got it